scripts/asm-installer/install_asm

```diff
@@ -0,0 +1,864 @@
+      MISSING_ARGS=1
+      warn "Missing value for ${REQUIRED_ARG}"
+    fi
+    readonly "${REQUIRED_ARG}"
+  done <<EOF
+CLUSTER_LOCATION
+CLUSTER_NAME
+PROJECT_ID
+MODE
+CA
+EOF
+
+  if [[ "${MISSING_ARGS}" -ne 0 ]]; then
+    fatal_with_usage "Missing one or more required options."
+  fi
+
+  while read -r FLAG; do
+    if [[ "${!FLAG}" -ne 0 && "${!FLAG}" -ne 1 ]]; then
+      fatal "${FLAG} must be 0 (off) or 1 (on) if set via environment variables."
+    fi
+    readonly "${FLAG}"
+  done <<EOF
+DRY_RUN
+ENABLE_APIS
+ONLY_VALIDATE
+VERBOSE
+EOF
+
+  if [[ -n "$SERVICE_ACCOUNT" && -z "$KEY_FILE" || -z "$SERVICE_ACCOUNT" && -n "$KEY_FILE" ]]; then
+    fatal "Service account and key file must be used together."
+  fi
+
+  # since we cd to a tmp directory, we need the absolute path for the key file
+  # and yaml file
+  if [[ -f "${KEY_FILE}" ]]; then
+    KEY_FILE="$(readlink -f "$KEY_FILE")"
+    readonly KEY_FILE
+  elif [[ -n "${KEY_FILE}" ]]; then
+    fatal "Couldn't find key file ${KEY_FILE}."
+  fi
+
+  if [[ -f "${OPERATOR_OVERLAY}" ]]; then
+    OPERATOR_OVERLAY="$(readlink -f "$OPERATOR_OVERLAY")"
+    readonly OPERATOR_OVERLAY
+  elif [[ -f "${OPERATOR_OVERLAY}" ]]; then
+    fatal "Couldn't find yaml file ${OPERATOR_OVERLAY}."
+  fi
+
+  set_kpt_package_url
+  WORKLOAD_POOL="${PROJECT_ID}.svc.id.goog"
+}
+
+set_kpt_package_url() {
+  local CA_OPT
+  CA_OPT=""
+  if [[ "${CA}" = "citadel" ]]; then
+    CA_OPT="-citadel"
+  fi
+  local URL_PREFIX
+  URL_PREFIX="https://github.com/GoogleCloudPlatform/anthos-service-mesh-packages"
+  KPT_GH_URL="${URL_PREFIX}/tree/release-1.6-asm/asm${CA_OPT}"
+  KPT_URL="${URL_PREFIX}.git/asm${CA_OPT}@release-1.6-asm"
+  readonly KPT_URL; readonly KPT_GH_URL;
+}
+
+auth_service_account() {
+  info "Authorizing ${SERVICE_ACCOUNT} with ${KEY_FILE}..."
+  run gcloud auth activate-service-account \
+    --project="${PROJECT_ID}" \
+    "${SERVICE_ACCOUNT}" \
+    --key-file="${KEY_FILE}"
+}
+
+#######
+# set_up_local_workspace does everything that the script needs to avoid
+# polluting the environment or current working directory
+#######
+set_up_local_workspace() {
+  info "Setting up a temporary workspace..."
+  WORKSPACE_DIR="$(mktemp -d)"
+  if [[ -z "${WORKSPACE_DIR}" ]]; then
+    fatal "Encountered error when running mktemp -d!"
+  fi
+  pushd "$WORKSPACE_DIR" > /dev/null
+
+  info "Generating a new kubeconfig..."
+  KUBECONFIG="$(mktemp)"
+  if [[ -z "${KUBECONFIG}" ]]; then
+    fatal "Encountered error when running mktemp!"
+  fi
+  export KUBECONFIG
+}
+
+cleanup_local_workspace() {
+  run rm "${KUBECONFIG}"
+  run popd
+  run rm -r "${WORKSPACE_DIR}"
+}
+
+### Environment validation functions ###
+validate_dependencies() {
+  validate_cli_dependencies
+  # configure kubectl does have side effects but we've generated a temprorary
+  # kubeconfig so we're not breaking the promise that --only_validate gives
+  configure_kubectl
+  validate_packages_fetchable
+  validate_gcp_resources
```

Comment:
validate_gcp_resources needs to happen before configure_kubectl to make sure cluster and project exist.